app/dao/notifications_dao.py

```diff
@@ -361,6 +364,54 @@ def _delete_notifications(
     return deleted
 
 
+def insert_update_notification_history(notification_type, date_to_delete_from, service_id):
+    notifications = db.session.query(
+        Notification.id,
+        Notification.job_id,
+        Notification.job_row_number,
+        Notification.service_id,
+        Notification.template_id,
+        Notification.template_version,
+        Notification.api_key_id,
+        Notification.key_type,
+        Notification.billable_units,
+        Notification.notification_type,
+        Notification.created_at,
+        Notification.sent_at,
+        Notification.sent_by,
+        Notification.updated_at,
+        Notification.status,
+        Notification.reference,
+        Notification.client_reference,
+        Notification.international,
+        Notification.phone_prefix,
+        Notification.rate_multiplier,
+        Notification.created_by_id,
+        Notification.postage
+    ).filter(
+        Notification.notification_type == notification_type,
+        Notification.service_id == service_id,
+        Notification.created_at < date_to_delete_from,
+        Notification.key_type != KEY_TYPE_TEST
+    ).all()
```

Comment:
rather than calling .all (and invoking big data serialisation and deserialisation costs), does an `insert into notification_history from (select ... from notifications)` query work?

https://docs.sqlalchemy.org/en/13/core/dml.html?highlight=from_select#sqlalchemy.sql.expression.Insert.from_select

Reply:
It doesn't look like it can handle the `on_conflict_do_update`